File: bindings/js/src/resolve.ts
/**
 * The core resolution algorithm — the cross-language contract.
 *
 * See `docs/architecture.md` and the Rust reference (`core-rs/src/resolve.rs`).
 */

import { data } from "./data.js";
import { sideNames, teamCode, teamName } from "./crosswalk.js";
import { normalize, teamMatches } from "./normalize.js";
import type {
  EspnGame,
  EspnSide,
  MapResult,
  Outcome,
  PmEvent,
  PmMarket,
  TeamBlock,
} from "./models.js";

function teamBlock(league: string, side: EspnSide): TeamBlock {
  return {
    espn_abbr: side.abbr,
    name: teamName(league, side),
    pm_code: teamCode(league, side.abbr),
  };
}

function isDraw(m: PmMarket): boolean {
  return m.slug.endsWith("-draw") || normalize(m.groupItemTitle).startsWith("draw");
}

function idxStr(arr: string[], i: number): string {
  return i >= 0 && i < arr.length ? arr[i] : "";
}

/** Resolve one ESPN game against a candidate Polymarket event. */
export function resolve(league: string, game: EspnGame, pm: PmEvent): MapResult {
  const homeBlock = teamBlock(league, game.home);
  const awayBlock = teamBlock(league, game.away);
  const kind = data().league(league)?.kind ?? "";

  const result: MapResult = {
    resolved: false,
    league,
    kind,
    espn_event_id: game.espnEventId,
    pm_event_slug: pm.slug,
    pm_market_slug: null,
    home: homeBlock,
    away: awayBlock,
    outcomes: [],
  };

  const homeNames = sideNames(league, game.home);
  const awayNames = sideNames(league, game.away);

  if (kind === "three_way") {
    let homeLeg: PmMarket | null = null;
    let awayLeg: PmMarket | null = null;
    let drawLeg: PmMarket | null = null;
    for (const m of pm.markets) {
      if (isDraw(m)) {
        if (drawLeg === null) drawLeg = m;
        continue;
      }
      const label = m.groupItemTitle === "" ? m.question : m.groupItemTitle;
      if (homeLeg === null && teamMatches(homeNames, label)) {
        homeLeg = m;
      } else if (awayLeg === null && teamMatches(awayNames, label)) {
        awayLeg = m;
      }
    }
    const legOutcome = (
      selection: "home" | "away" | "draw",
      team: string | null,
      m: PmMarket,
    ): Outcome => ({
      selection,
      team,
      pm_market_slug: m.slug,
      pm_outcome: idxStr(m.outcomes, 0),
      outcome_index: 0,
      token_id: idxStr(m.clobTokenIds, 0),
      price: idxStr(m.outcomePrices, 0),
    });
    if (homeLeg !== null) {
      result.outcomes.push(legOutcome("home", result.home.name, homeLeg));
    }
    if (awayLeg !== null) {
      result.outcomes.push(legOutcome("away", result.away.name, awayLeg));
    }
    if (drawLeg !== null) {
      result.outcomes.push(legOutcome("draw", null, drawLeg));
    }
    result.resolved = homeLeg !== null && awayLeg !== null;
    return result;
  }

  // two_way
  let ml = pm.markets.find((m) => m.slug === pm.slug);
  if (ml === undefined) {
    ml = pm.markets.find((m) => m.sportsMarketType === "moneyline");
  }
  if (ml === undefined) {
    return result;
  }
  result.pm_market_slug = ml.slug;

  let homeIdx: number | null = null;
  let awayIdx: number | null = null;
  for (let i = 0; i < ml.outcomes.length; i++) {
    const o = ml.outcomes[i];
    if (homeIdx === null && teamMatches(homeNames, o)) {
      homeIdx = i;
    } else if (awayIdx === null && teamMatches(awayNames, o)) {
      awayIdx = i;
    }
  }
  if (homeIdx !== null && awayIdx !== null && homeIdx !== awayIdx) {
    for (let i = 0; i < ml.outcomes.length; i++) {
      let selection: "home" | "away";
      let team: string;
      if (i === homeIdx) {
        selection = "home";
        team = result.home.name;
      } else if (i === awayIdx) {
        selection = "away";
        team = result.away.name;
      } else {
        continue;
      }
      result.outcomes.push({
        selection,
        team,
        pm_market_slug: ml.slug,
        pm_outcome: idxStr(ml.outcomes, i),
        outcome_index: i,
        token_id: idxStr(ml.clobTokenIds, i),
        price: idxStr(ml.outcomePrices, i),
      });
    }
    result.resolved = true;
  }
  return result;
}

File: core-rs/src/resolve.rs
//! The core resolution algorithm — the cross-language contract (see `docs/architecture.md`).

use crate::crosswalk::{side_names, team_code, team_name};
use crate::data::data;
use crate::models::{EspnGame, MapResult, Outcome, PmEvent, PmMarket, TeamBlock};
use crate::normalize::{normalize, team_matches};

fn team_block(league: &str, side: &crate::models::EspnSide) -> TeamBlock {
    TeamBlock {
        espn_abbr: side.abbr.clone(),
        name: team_name(league, side),
        pm_code: team_code(league, &side.abbr),
    }
}

fn is_draw(m: &PmMarket) -> bool {
    m.slug.ends_with("-draw") || normalize(&m.group_item_title).starts_with("draw")
}

fn idx_str(arr: &[String], i: usize) -> String {
    arr.get(i).cloned().unwrap_or_default()
}

/// Resolve one ESPN game against a candidate Polymarket event.
pub fn resolve(league: &str, game: &EspnGame, pm: &PmEvent) -> MapResult {
    let home_block = team_block(league, &game.home);
    let away_block = team_block(league, &game.away);
    let kind = data()
        .league(league)
        .map(|c| c.kind.clone())
        .unwrap_or_default();

    let mut result = MapResult {
        resolved: false,
        league: league.to_string(),
        kind: kind.clone(),
        espn_event_id: game.espn_event_id.clone(),
        pm_event_slug: pm.slug.clone(),
        pm_market_slug: None,
        home: home_block,
        away: away_block,
        outcomes: Vec::new(),
    };

    let home_names = side_names(league, &game.home);
    let away_names = side_names(league, &game.away);

    if kind == "three_way" {
        let mut home_leg = None;
        let mut away_leg = None;
        let mut draw_leg = None;
        for m in &pm.markets {
            if is_draw(m) {
                draw_leg.get_or_insert(m);
                continue;
            }
            let label = if m.group_item_title.is_empty() {
                &m.question
            } else {
                &m.group_item_title
            };
            if home_leg.is_none() && team_matches(&home_names, label) {
                home_leg = Some(m);
            } else if away_leg.is_none() && team_matches(&away_names, label) {
                away_leg = Some(m);
            }
        }
        let leg_outcome = |selection: &str, team: Option<String>, m: &PmMarket| Outcome {
            selection: selection.to_string(),
            team,
            pm_market_slug: m.slug.clone(),
            pm_outcome: idx_str(&m.outcomes, 0),
            outcome_index: 0,
            token_id: idx_str(&m.clob_token_ids, 0),
            price: idx_str(&m.outcome_prices, 0),
        };
        if let Some(m) = home_leg {
            result
                .outcomes
                .push(leg_outcome("home", Some(result.home.name.clone()), m));
        }
        if let Some(m) = away_leg {
            result
                .outcomes
                .push(leg_outcome("away", Some(result.away.name.clone()), m));
        }
        if let Some(m) = draw_leg {
            result.outcomes.push(leg_outcome("draw", None, m));
        }
        result.resolved = home_leg.is_some() && away_leg.is_some();
        return result;
    }

    // two_way
    let ml = pm.markets.iter().find(|m| m.slug == pm.slug).or_else(|| {
        pm.markets
            .iter()
            .find(|m| m.sports_market_type.as_deref() == Some("moneyline"))
    });
    let Some(ml) = ml else {
        return result;
    };
    result.pm_market_slug = Some(ml.slug.clone());

    let mut home_idx = None;
    let mut away_idx = None;
    for (i, o) in ml.outcomes.iter().enumerate() {
        if home_idx.is_none() && team_matches(&home_names, o) {
            home_idx = Some(i);
        } else if away_idx.is_none() && team_matches(&away_names, o) {
            away_idx = Some(i);
        }
    }
    if let (Some(hi), Some(ai)) = (home_idx, away_idx) {
        if hi != ai {
            for i in 0..ml.outcomes.len() {
                let (selection, team) = if i == hi {
                    ("home", Some(result.home.name.clone()))
                } else if i == ai {
                    ("away", Some(result.away.name.clone()))
                } else {
                    continue;
                };
                result.outcomes.push(Outcome {
                    selection: selection.to_string(),
                    team,
                    pm_market_slug: ml.slug.clone(),
                    pm_outcome: idx_str(&ml.outcomes, i),
                    outcome_index: i as u32,
                    token_id: idx_str(&ml.clob_token_ids, i),
                    price: idx_str(&ml.outcome_prices, i),
                });
            }
            result.resolved = true;
        }
    }
    result
}
